![fa-linux](http://fa2png.io/media/icons/font-awesome/4-7-0/linux/20/0/2c3e50_none.png) 
[![Travis-CI Build Status](https://travis-ci.org/waldronlab/GSEABenchmarkeR.svg?branch=master)](https://travis-ci.org/waldronlab/GSEABenchmarkeR)
![fa-windows](http://fa2png.io/media/icons/font-awesome/4-7-0/windows/20/0/2c3e50_none.png)
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/waldronlab/GSEABenchmarkeR?branch=master&svg=true)](https://ci.appveyor.com/project/waldronlab/GSEABenchmarkeR)


# GSEABenchmarkeR

The _GSEABenchmarkeR_ package implements an extendable framework for the 
reproducible evaluation of set- and network-based methods for the enrichment 
analysis of gene expression data. This includes support for the efficient 
execution of these methods on comprehensive real data compendia (microarray and 
RNA-seq) using parallel computation on standard workstations as well as 
institutional computer grids. Methods can then be assessed with respect to 
runtime, statistical significance, and relevance of the results for the 
investigated phenotypes.
    
## Installation

You can install GSEABenchmarkeR from github with:


``` r
# install.packages("devtools")
devtools::install_github("waldronlab/GSEABenchmarkeR")
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
## basic example code
```
